src/core/server/saved_objects/service/lib/get_references_deep.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { ISavedObjectTypeRegistry } from '../../saved_objects_type_registry';
+import type { SavedObjectsSerializer } from '../../serialization';
+import type { SavedObject } from '../../types';
+import type {
+  SavedObjectReferenceWithContext,
+  SavedObjectsGetReferencesDeepObject,
+  SavedObjectsGetReferencesDeepResponse,
+} from '../saved_objects_client';
+import { getRootFields } from './included_fields';
+import type { RepositoryEsClient } from './repository_es_client';
+import { getSavedObjectFromSource } from './utils';
+
+export interface GetReferencesDeepParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: RepositoryEsClient;
+  serializer: SavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  objects: SavedObjectsGetReferencesDeepObject[];
+}
+
+/**
+ * Gets all references and transitive references of the given objects. Ignores any object and/or reference that is not a multi-namespace
+ * type.
+ */
+export async function getReferencesDeep({
+  registry,
+  allowedTypes,
+  client,
+  serializer,
+  getIndexForType,
+  objects,
+}: GetReferencesDeepParams): Promise<SavedObjectsGetReferencesDeepResponse> {
+  const inboundReferencesMap = objects.reduce(
+    // Add the input objects to the references map so they are returned with the results, even if they have no inbound references
+    (acc, { type, id }) => acc.set(`${type}:${id}`, new Map()),
+    new Map<string, Map<string, string>>()
+  );
+  const objectMap = new Map<string, SavedObject | null>();
+
+  const rootFields = getRootFields();
+  const makeBulkGetDocs = (objectsToGet: SavedObjectsGetReferencesDeepObject[]) =>
+    objectsToGet.map(({ type, id }) => ({
+      _id: serializer.generateRawId(undefined, type, id),
+      _index: getIndexForType(type),
+      _source: rootFields, // Optimized to only retrieve root fields (ignoring type-specific fields)
+    }));
+  const objectFilter = ({ type }: SavedObjectsGetReferencesDeepObject) =>
+    allowedTypes.includes(type) && registry.isMultiNamespace(type);
```

Comment:
Two reasons:
1. Because of the way that we keep track of object keys via the `inboundReferencesMap` and the `objectsMap`, using the same approach without modifications could result in collisions for single-namespace objects in different spaces.
2. Retrieving the referenced objects gets hairy if you have a multi-namespace object type that has a reference to a single-namespace object type -- then you don't know what namespace to check in. It _shouldn't_ happen, but we don't have any guard rails around references.

This approach satisfies our immediate needs and is less complex, so it seemed safer :)

Reply:
I figured it was a simplification measure, but we should make sure that we're OK not supporting this scenario. For example, when we produce The List™️ of saved object types to convert in 8.0, we need to make sure that those objects can't produce a reference graph that includes a saved object type that we do not intend to convert. Or if it can produce such a graph, we need to make it very clear that this function will silently omit single-namespace types from the response.

If we move forward with the current approach, then I think we should brainstorm other names for this function, as `getReferencesDeep` does not convey the current limitations. Some bad ideas include:
- `collectMultiNamespaceReferences`
- `getMultiNamespaceReferencesDeep`
- `collectSharableReferences` / `getSharableReferencesDeep` (I like this if we didn't already use the term "sharable" to mean [something else](https://github.com/elastic/kibana/blob/5c3c3efdd87089fb1a326854c83397a7253bd7c6/src/core/server/saved_objects/saved_objects_type_registry.ts#L101))